File: cdk_lib/stack.py
import hashlib
import inspect
import os
import subprocess
from pathlib import Path

from aws_cdk import core, aws_lambda


class PythonStack(core.Stack):

    def __init__(self, scope: core.Construct, construct_id: str, **kwargs) -> None:
        self.project_dir = os.path.dirname(inspect.getfile(self.__class__))
        super().__init__(scope, construct_id, **kwargs)
        self.lambda_layers = [self.create_dependencies_layer()]

    def create_dependencies_layer(self) -> aws_lambda.LayerVersion:
        layers_dir = Path(self.project_dir, '.layers')
        layers_dir.mkdir(exist_ok=True)
        requirements = subprocess.check_output(
            f'poetry export --with-credentials --without-hashes'.split()
        ).decode("utf-8")
        requirements = '\n'.join([line for line in requirements.splitlines() if not line.startswith('boto')])

        requirements_hash = hashlib.sha256(requirements.encode('utf-8')).hexdigest()
        destination_dir = layers_dir.joinpath(requirements_hash)
        if not destination_dir.exists():
            destination_dir.mkdir()
            requirements_file = destination_dir.joinpath('requirements.txt')
            with requirements_file.open('w') as f:
                f.write(requirements)
            print('running pip install...')
            subprocess.check_call(
                f'pip install --no-deps -r {requirements_file} -t {destination_dir}/python'.split())

        layer_id = f'{self.stack_name}-dependencies'
        layer_code = aws_lambda.Code.from_asset(str(destination_dir))

        return aws_lambda.LayerVersion(self, layer_id, code=layer_code)

    def code_asset(self, relative_path):
        """Turn a path relative to the project dir into a lambda code asset.
        This makes it possible to test stack creation from within the test dir"""
        return aws_lambda.Code.from_asset(os.path.join(self.project_dir, relative_path))

File: cdk/lib/common/stack.ts
import {Construct, Stack, StackProps} from "@aws-cdk/core"
import {AssetCode, Code, LayerVersion} from "@aws-cdk/aws-lambda"
import * as fs from "fs"
import * as child_process from "child_process"
import * as path from "path"
import * as crypto from "crypto"

export class PythonStack extends Stack {
    private readonly projectDir: string
    public lambdaLayers: LayerVersion[]


    constructor(scope: Construct, constructId: string, props?: StackProps) {
        super(scope, constructId, props)

        this.projectDir = path.join('.', '..')
        this.lambdaLayers = [this.createDependenciesLayer()]
    }

    createDependenciesLayer(): LayerVersion {
        const layersDir = path.join(this.projectDir, '.layers')

        if (!fs.existsSync(layersDir)) {
            fs.mkdirSync(layersDir)
        }

        const stdout = child_process.execSync('../pw poetry export --with-credentials --without-hashes', {encoding: 'utf8'})
        let requirementsArray = stdout.match(/[^\r\n]+/g) || []
        requirementsArray = requirementsArray.filter(line => !line.startsWith('boto'))
        const requirements = requirementsArray.join('\n')
        const requirementsHash = crypto.createHash('md5').update(requirements).digest('hex')
        const destinationDir = path.join(layersDir, requirementsHash)
        const requirementsFile = path.join(destinationDir, 'requirements.txt')
        const pythonDir = path.join(destinationDir, 'python')

        if (!fs.existsSync(destinationDir)) {
            fs.mkdirSync(destinationDir)
        }
        if (!fs.existsSync(requirementsFile)) {
            fs.writeFileSync(requirementsFile, requirements)
        }
        if (!fs.existsSync(pythonDir)) {
            console.info('running pip install...')
            const result = child_process.spawnSync('pip3',
                ['install', '--no-deps', '-r', requirementsFile.toString(), '-t', pythonDir.toString()])
            if (result.status) {
                throw new Error(`pip3 install exited with non-zero code: ${result.status}`)
            }
        }
        const layerId = `${this.stackName}-dependencies`
        const layerCode = Code.fromAsset(destinationDir)
        return new LayerVersion(this, layerId, {code: layerCode})
    }

    createCodeAsset(relativePath: string): AssetCode {
        return Code.fromAsset(path.join(this.projectDir, relativePath))
    }
}
